### PR TITLE
feat!: rewrite to TS and adjust to the v2 ParserJS

### PR DESCRIPTION
**Pull request by magicmatatjahu**
**Description**

- Rewrite to TS.
- Adjust exported module to the new v2 ParserJS Custom Schema Parser interface - new `validate()` function.
- Improve testing and move to the Jest.
- Update Readme.
- Add @magicmatatjahu to the codeowners.

**Related issue(s)**
Resolves https://github.com/asyncapi/openapi-schema-parser/issues/140

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  let result: SchemaValidateResult[] = [];
+  const valid = validator(input.data);
+  if (!valid && validator.errors) {
+    result = ajvToSpectralResult(input.path, [...validator.errors]);
+  }
+
+  return result;
+}
+
+async function parse(input: ParseSchemaInput<unknown, unknown>): Promise<v2.SchemaObject> {
```
Comment (smoya) [inline src/index.ts:34]:
```suggestion
async function parse(input: ParseSchemaInput<unknown, unknown>): Promise<AsyncAPISchema> {
```

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
+
+import Ajv from 'ajv';
+import addFormats from 'ajv-formats';
+import ajvErrors from 'ajv-errors';
+import { jsonSchemaV3 } from './json-schema-v3';
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const toJsonSchema = require('@openapi-contrib/openapi-schema-to-json-schema');
+
+import type { SchemaParser, ValidateSchemaInput, ParseSchemaInput, SchemaValidateResult } from '@asyncapi/parser';
```
Comment (smoya) [inline src/index.ts:9]:
```suggestion
import type { SchemaParser, ValidateSchemaInput, ParseSchemaInput, SchemaValidateResult, AsyncAPISchema } from '@asyncapi/parser';
```

File: src/index.ts (hunk 1 above)
Comment (magicmatatjahu) [reply src/index.ts:34]: I know it perfectly well 😂 I wait for your PR.

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import type { v2 } from '@asyncapi/parser/esm/spec-types';
```
Comment (smoya) [inline src/index.ts:10]:
```suggestion
```

File: src/index.ts (hunk 1 above)
Comment (smoya) [reply src/index.ts:34]: self service! https://github.com/asyncapi/parser-js/pull/656 :D

**Comment by smoya** [comment]
@magicmatatjahu I think using `next-major` as target branch its a better approach, so this can be merged and released as beta first.

File: CODEOWNERS [hunk 4]
```diff
@@ -5,4 +5,4 @@
 # For more details, read the following article on GitHub: https://help.github.com/articles/about-codeowners/.
 
 # The default owners are automatically added as reviewers when you open a pull request unless different owners are specified in the file.
-* @fmvilas @derberg @asyncapi-bot-eve
+* @fmvilas @derberg @magicmatatjahu @asyncapi-bot-eve
```
Comment (fmvilas) [inline CODEOWNERS:8]: I think this should go in a separate PR 🙏 

```suggestion
* @fmvilas @derberg @asyncapi-bot-eve
```

**Review by fmvilas — CHANGES_REQUESTED** [review]
LGTM in general. Just added a comment.

File: CODEOWNERS (hunk 4 above)
Comment (magicmatatjahu) [reply CODEOWNERS:8]: Ok :)

**Review by fmvilas — DISMISSED** [review]

**Comment by derberg** [comment]
I have same question as https://github.com/asyncapi/avro-schema-parser/pull/168#pullrequestreview-1171929231, but this PR is a bit different and points to `master` directly. Should we really release it as major and automatically bump it in `generator`?

**Comment by magicmatatjahu** [comment]
@derberg 

> Should we really release it as major and automatically bump it in generator?

We will but it doesn't work with parser in generator. In this PR I made integration https://github.com/asyncapi/generator/pull/833 but I'm fighting with snapshots 😆  

> but this PR is a bit different and points to master directly.

Yeah, because I didn't have `next-major` in this repo :)

**Comment by smoya** [comment]
relevant

**Comment by smoya** [comment]
@fmvilas @derberg can any of you create the `next-major` branch? I want to move forward with this PR. Once is done, @magicmatatjahu would u mind changing the base branch and update it? 

Thanks

**Review by smoya — CHANGES_REQUESTED** [review]
See https://github.com/asyncapi/openapi-schema-parser/pull/142/files#r1153125770

**Comment by smoya** [comment]
@fmvilas @derberg please approve when you have a chance. Last change was just the fix on dependencies

**Review by smoya — APPROVED** [review]
